File: Anul 2, Sem 1/MAP/Lab 7 C#/Repository/InMemoryRepository.cs
using System;
using System.Collections.Generic;
using System.Text;
using Lab7.Validator;
using Lab7.Domain;
using System.Linq;

namespace Lab7.Repository
{
    class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
    {

        protected IDictionary<ID, E> entities = new Dictionary<ID, E>();

        public E Delete(ID id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<E> FindAll()
        {
            return entities.Values.ToList<E>();
        }

        public E FindOne(ID id)
        {
            return entities[id];
        }

        public E Save(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("entity must not be null");
            if (this.entities.ContainsKey(entity.ID))
            {
                return entity;
            }
            this.entities[entity.ID] = entity;
            return default(E);
        }

        public E Update(E entity)
        {
            throw new NotImplementedException();
        }
    }
}

File: Anul 2, Sem 1/MAP/proiect-lab-schelet/src/main/java/socialnetwork/repository/memory/InMemoryRepository.java
package socialnetwork.repository.memory;

import socialnetwork.domain.Entity;
import socialnetwork.domain.validators.ValidationException;
import socialnetwork.domain.validators.Validator;
import socialnetwork.repository.RepoException;
import socialnetwork.repository.Repository;

import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID,E> {

    private Validator<E> validator;
    protected Map<ID,E> entities;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities=new HashMap<ID,E>();
    }

    @Override
    public E findOne(ID id){
        if (id==null)
            throw new IllegalArgumentException("Id must be not null\n");
        return entities.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        if (entity==null)
            throw new IllegalArgumentException("Entity must be not null\n");
        validator.validate(entity);

        if(entities.get(entity.getId()) != null)
            throw new RepoException("Id existent!\n");

        entities.put(entity.getId(),entity);

        return null;
    }

    @Override
    public E delete(ID id)
    {
        if(id == null)
            throw new IllegalArgumentException("Id-ul nu poate sa fie null!\n");

        return entities.remove(id);
    }

    @Override
    public E update(E entity) {

        if(entity == null)
            throw new IllegalArgumentException("entity must be not null!\n");
        validator.validate(entity);

        entities.put(entity.getId(),entity);

        if(entities.get(entity.getId()) != null) {
            entities.put(entity.getId(),entity);
            return null;
        }
        return entity;

    }

}
